src/metabase/models/card.clj

```diff
@@ -78,4 +80,44 @@
   revision/IRevisioned
   {:serialize-instance serialize-instance
    :revert-to-revision revision/default-revert-to-revision
-   :describe-diff      revision/default-describe-diff})
+   :diff-map           revision/default-diff-map
+   :diff-str           revision/default-diff-str})
+
+
+;;; ## ---------------------------------------- DEPENDENCIES ----------------------------------------
+
+
+(defn- parse-filter-subclause [subclause]
+  (match subclause
+    ["SEGMENT" (segment-id :guard integer?)] segment-id
+    subclause                                nil))
+
+(defn- parse-filter [clause]
+  (match clause
+    ["AND" & subclauses] (mapv parse-filter subclauses)
+    ["OR" & subclauses]  (mapv parse-filter subclauses)
+    subclause            (parse-filter-subclause subclause)))
+
+(defn- extract-segment-ids [filter-clause]
+  (when filter-clause
+    (->> (parse-filter filter-clause)
+         flatten
+         (filter identity)
+         set)))
+
+(defn- extract-metric-ids [aggregation-clause]
+  (match aggregation-clause
+    ["METRIC" (metric-id :guard integer?)] #{metric-id}
```

Comment:
yeah, i could go either way here.  What happened is that i added `SEGMENT` first, which was uppercased to match the various filter clauses.  then when i went to add `METRIC` it was either a choice between staying consistent between the macro type clauses that i was creating (uppercased), or following the convention of the other aggregation clauses (lowercased).

I figured it was better to try and make all the macros consistent, so that's why i used uppercase.

Happy to switch it to lowercase if you think it's better that way.